src/hotspot/share/oops/oop.inline.hpp

```diff
@@ -266,28 +267,78 @@ bool oopDesc::is_forwarded() const {
 
 // Used by scavengers
 void oopDesc::forward_to(oop p) {
+  assert(p != cast_to_oop(this) || !UseAltGCForwarding, "Must not be called with self-forwarding");
   markWord m = markWord::encode_pointer_as_mark(p);
-  assert(m.decode_pointer() == p, "encoding must be reversible");
+  assert(forwardee(m) == p, "encoding must be reversible");
   set_mark(m);
 }
 
+void oopDesc::forward_failed() {
```

Comment:
Yeah.
It's even more confusing that we now have the notion of forward-failed, which aims to hide the implementation detail of self-forwarding, but forwardee() still exposes it. And probably has to, because that is how the forwarding logic of GCs currently work, and I'm not sure it is useful to change that. I need to mull over this a bit.